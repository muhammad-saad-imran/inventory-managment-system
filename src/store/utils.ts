import { completeLoading, startLoading } from "@/store/features/loading";
import { createAsyncThunk, GetThunkAPI } from "@reduxjs/toolkit";

export const getThunkCreator = <T>(
  thunkName: string,
  getEntity: (id: string) => Promise<T>
) =>
  createAsyncThunk(
    thunkName,
    async (id: string, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        const entity = await getEntity(id);
        dispatch(completeLoading());
        return entity;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );

export const getAllThunkCreator = <T>(
  thunkName: string,
  getAllEntity: (input: string) => Promise<T[]>
) =>
  createAsyncThunk(
    thunkName,
    async (name: string, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        const entity = await getAllEntity(name);
        dispatch(completeLoading());
        return entity;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );

export const mutationThunkCreator = <T>(
  thunkName: string,
  mutateEntity: (values: T) => Promise<T>
) =>
  createAsyncThunk(
    thunkName,
    async (values: T, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        const entity = await mutateEntity(values);
        dispatch(completeLoading());
        return entity;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );

export const deleteThunkCreator = (
  thunkName: string,
  deleteEntity: (id: string) => Promise<void>
) =>
  createAsyncThunk(
    thunkName,
    async (id: string, { dispatch, rejectWithValue }) => {
      try {
        dispatch(startLoading());
        await deleteEntity(id);
        dispatch(completeLoading());
        return;
      } catch (error) {
        dispatch(completeLoading());
        alert("Error fetching products");
        rejectWithValue(error);
      }
    }
  );
