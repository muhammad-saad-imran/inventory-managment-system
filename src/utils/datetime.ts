import dayjs from "dayjs";
import utc from "dayjs/plugin/utc";

dayjs.extend(utc);

type DateInput = {
  date?: dayjs.ConfigType;
  inputDate?: string;
  outputDate: string;
};

export const formatDate = ({ date, inputDate, outputDate }: DateInput) => {
  return dayjs(date, inputDate).format(outputDate);
};

export const formatPrice = (price: number) =>
  new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  }).format(price);
