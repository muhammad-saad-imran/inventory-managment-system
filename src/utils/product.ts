import { SupabaseClient } from "@supabase/supabase-js";
import { DB_TABLES } from "@/utils/supabase/types";

export const getProduct = async (id: string, supabase: SupabaseClient) => {
  try {
    const { data: product, error } = await supabase
      .from(DB_TABLES.PRODUCTS)
      .select()
      .eq("id", id)
      .limit(1)
      .single();

    if (error) throw error;

    return product;
  } catch (error) {
    throw error;
  }
};

export const updateProductStock = async (
  id: string,
  stock_quantity: number,
  supabase: SupabaseClient
) => {
  try {
    const { error } = await supabase
      .from(DB_TABLES.PRODUCTS)
      .update({ stock_quantity })
      .eq("id", id);

    if (error) throw error;
  } catch (error) {
    throw error;
  }
};
