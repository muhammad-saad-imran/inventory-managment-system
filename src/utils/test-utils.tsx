import { PropsWithChildren } from 'react'
import { Provider } from 'react-redux'
import { render, type RenderOptions } from '@testing-library/react'

import { AppStore, makeStore } from '@/store'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    // Automatically create a store instance if no store was passed in
    store = makeStore(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }  
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}