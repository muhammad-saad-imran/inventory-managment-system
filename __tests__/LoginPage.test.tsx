import LoginPage from "@/app/page";
import { renderWithProviders } from "@/utils/test-utils";

jest.mock("@/utils/supabase/client");

describe("Login Page", () => {
  it("should match snapshot", () => {
    const { container } = renderWithProviders(<LoginPage />);
    expect(container).toMatchSnapshot();
  });
});
