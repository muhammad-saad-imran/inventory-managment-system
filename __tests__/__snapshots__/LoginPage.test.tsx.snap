// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should match snapshot 1`] = `
<div>
  <div
    class="sc-267e98d2-0 eAqqCL"
  >
    <div
      class="flex p-5 bg-white rounded-lg shadow-lg"
    >
      <img
        alt=""
        class="pr-8"
        data-nimg="1"
        decoding="async"
        height="500"
        loading="lazy"
        src="/_next/image?url=%2Flogin.jpg&w=1080&q=75"
        srcset="/_next/image?url=%2Flogin.jpg&w=640&q=75 1x, /_next/image?url=%2Flogin.jpg&w=1080&q=75 2x"
        style="color: transparent;"
        width="500"
      />
      <form
        class="flex flex-col justify-center w-96 p-8 gap-6"
        method="POST"
      >
        <div
          class="flex flex-col gap-1 w-full"
        >
          <label
            for="email"
          >
            Email:
          </label>
          <input
            class="sc-23ee340d-0 sc-23ee340d-1 cwYaeo jqzeAh"
            id="email"
            name="email"
            required=""
            type="email"
            value=""
          />
        </div>
        <div
          class="flex flex-col gap-1 w-full"
        >
          <label
            for="password"
          >
            Password:
          </label>
          <input
            class="sc-23ee340d-0 sc-23ee340d-1 cwYaeo jqzeAh"
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
        </div>
        <button
          class="sc-7d679c72-0 zDViG ml-auto"
          type="submit"
        >
          Log in
        </button>
      </form>
    </div>
  </div>
</div>
`;
